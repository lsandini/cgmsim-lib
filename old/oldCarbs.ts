import moment = require('moment');
import logger from '../src/utils';

export const oldCarbs = (meal, carbAbsTime) => {
  const meals = meal.map((entry) => ({
    ...entry,
    time: moment(entry.time).valueOf(),
  }));
  const timeSinceMealMin = meals.map((entry) => ({
    ...entry,
    mills: entry.time,
    time: (Date.now() - moment(entry.time).valueOf()) / (1000 * 60),
  }));
  logger.debug(
    'this is the trimmed down meals and time since last meal:%o',
    timeSinceMealMin
  );

  // this is for the calculations of carbs ingestion
  let lastMeals = timeSinceMealMin.filter(function (e) {
    return e.time <= 360; // keep only the meals from the last 6 hours or 360 min
  });

  var karbs = lastMeals;

  const fast_carbAbsTime = carbAbsTime / 6; // = 1 h or 60 min
  const slow_carbAbsTime = carbAbsTime / 1.5; // = 4 h or 240 min

  let timeSinceMealAct = karbs.map((entry) => {
    var t = entry.time;
    var carbs_g = entry.carbs;

    // the first 40g of every meal are always considered fast carbs
    var fast = Math.min(entry.carbs, 40);

    // the amount exceeding 40 grams will be randomly split into fast and slow carbs
    var rest = entry.carbs - fast;
    var FSR = Math.random() * (0.4 - 0.1) + 0.1; // FSR = FAST RANDOM RATIO

    // all fast carbs counted together
    var fast_carbs = fast + FSR * rest;

    // the remainder is slow carbs
    var slow_carbs = (1 - FSR) * rest;

    //console.log(fast_random_ratio);
    //console.log(fast_carbs);
    //console.log(slow_carbs);
    logger.debug(
      'carbs_g:%o',
      carbs_g,
      'fast:%o',
      fast,
      'rest:%o',
      rest,
      'fast_carbs:%o',
      fast_carbs,
      'slow_carbs:%o',
      slow_carbs
    );

    if (t < fast_carbAbsTime / 2) {
      var AT2 = Math.pow(fast_carbAbsTime, 2);
      var fast_carbrate = (fast_carbs * 4 * t) / AT2;
      //	var COB = (fast_carbs * 2 * Math.pow(t, 2)) / AT2;
    } else if (t < fast_carbAbsTime) {
      var fast_carbrate =
        ((fast_carbs * 4) / fast_carbAbsTime) * (1 - t / fast_carbAbsTime);
      // var AAA = (4 * fast_carbs / fast_carbAbsTime);
      // var BBB = Math.pow(t, 2) / (2 * fast_carbAbsTime);
      //	var COB = (AAA * (t - BBB)) - fast_carbs;
    } else {
      var fast_carbrate = 0;
      //var COB = 0;
      logger.debug('fast carb absorption rate:%o', fast_carbrate);
    }

    if (t < slow_carbAbsTime / 2) {
      var AT2 = Math.pow(slow_carbAbsTime, 2);
      var slow_carbrate = (slow_carbs * 4 * t) / AT2;
      // var COB = (slow_carbs * 2 * Math.pow(t, 2)) / AT2;
    } else if (t < slow_carbAbsTime) {
      var slow_carbrate =
        ((slow_carbs * 4) / slow_carbAbsTime) * (1 - t / slow_carbAbsTime);
      // var AAA = (4 * slow_carbs / slow_carbAbsTime);
      // var BBB = Math.pow(t, 2) / (2 * slow_carbAbsTime);
      // var COB = (AAA * (t - BBB)) - slow_carbs;
    } else {
      var slow_carbrate = 0;
      // var COB = 0;
      logger.debug('slow carb absorption rate:%o', slow_carbrate);
    }

    return {
      ...entry,
      time: t,
      fast_carbrate: fast_carbrate,
      slow_carbrate: slow_carbrate,
      all_carbrate: fast_carbrate + slow_carbrate,
    };
  });
  logger.debug(timeSinceMealAct);

  var totalCarbRate = timeSinceMealAct.reduce(function (tot, arr) {
    return tot + arr.all_carbrate;
  }, 0);

  logger.debug(totalCarbRate);
  return totalCarbRate;

  // let carbrateString = JSON.stringify(totalCarbRate);
  // const fs = require('fs');
  // fs.writeFile("./files/latest_carbs.json", carbrateString, function (err, result) {
  // 	if (err) console.log('error', err);
  // });
};
