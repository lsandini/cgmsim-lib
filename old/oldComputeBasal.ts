import getLogger from '../src/utils';

const moment = require('moment');
export const oldComputeBasal = ({ entries }) => {
  var json = JSON.stringify(entries);
  var notes = JSON.parse(json);
  //getLogger().debug(notes);

  let basal = notes
    .filter((e) => e.notes)
    .map((e) => ({
      time: e.created_at,
      notes: e.notes,
      empty_space: e.notes.indexOf(' '),
    }));
  getLogger().debug('this is the filtered treatments (basal):%o', basal);
  getLogger().debug('length%o', basal.length); // returns the number of boluses or lenghth of the array

  let basals = basal.map((entry) => ({
    ...entry,
    time: moment(entry.time).valueOf(),
  }));
  let timeSinceBasalMin = basals.map((entry) => ({
    ...entry,
    time: (Date.now() - moment(entry.time).valueOf()) / (1000 * 60 * 60),
    drug: entry.notes.slice(0, 3),
    dose: parseInt(entry.notes.slice(entry.empty_space), 10),
  }));
  getLogger().debug(
    'this is the trimmed down insulin and time since injection data:%o',
    timeSinceBasalMin,
  );

  let lastBasals = timeSinceBasalMin.filter(function (e) {
    return e.time <= 45; // keep only the basals from the last 45 hours
  });
  getLogger().debug('these are the last basals:%o ', lastBasals);

  let lastGLA = lastBasals.filter(function (e) {
    return (
      e.drug === 'gla' ||
      e.drug === 'Gla' ||
      e.drug === 'lan' ||
      e.drug === 'Lan'
    ); // keep only the glas from the last 45 hours
  });
  getLogger().debug('these are the last glargines: %o', lastGLA);

  let lastDET = lastBasals.filter(function (e) {
    return (
      e.drug === 'det' ||
      e.drug === 'Det' ||
      e.drug === 'lev' ||
      e.drug === 'Lev'
    ); // keep only the dets from the last 45 hours
  });
  getLogger().debug('these are the last detemirs: %o', lastDET);

  let lastTOU = lastBasals.filter(function (e) {
    return e.drug === 'tou' || e.drug === 'Tou'; // keep only the tous from the last 45 hours
  });
  getLogger().debug('these are the last toujeos: %o', lastTOU);

  let lastDEG = lastBasals.filter(function (e) {
    return (
      e.drug === 'deg' ||
      e.drug === 'Deg' ||
      e.drug === 'tre' ||
      e.drug === 'Tre'
    ); // keep only the degs from the last 45 hours
  });
  getLogger().debug('these are the last degludecs: %o', lastDEG);

  const datadet = JSON.stringify(lastDET, null, 4);
  const datagla = JSON.stringify(lastGLA, null, 4);
  const datatou = JSON.stringify(lastTOU, null, 4);
  const datadeg = JSON.stringify(lastDEG, null, 4);

  return {
    lastDET,
    lastGLA,
    lastTOU,
    lastDEG,
  };
};
