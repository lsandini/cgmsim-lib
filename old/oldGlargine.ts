import logger from "../src/utils";

export default (weight, glargines) => {
	var resultGlaAct = 0;

	// const weight = parseInt(process.env.WEIGHT);
	// var moment = require('moment');
	// const glargines = require('./files/last_glargine.json');
	var jsongla = JSON.stringify(glargines);
	var glargine_data = JSON.parse(jsongla);
	logger.debug(glargine_data);

	// activities be expressed as U/min !!!
	let timeSinceGlargineAct = glargine_data.map(entry => {

		var time = entry.time;
		var dose = entry.dose;
		var duration = (22 + (12 * dose / weight));
		var peak = (duration / 2.5);
		var tp = peak;
		var td = duration;

		var tau = tp * (1 - tp / td) / (1 - 2 * tp / td);
		var a = 2 * tau / td;
		var S = 1 / (1 - a + (1 + a) * Math.exp(-td / tau));

		var glargineActivity = 0;
		return {
			...entry,
			time: time,
			glargineActivity: (dose * (S / Math.pow(tau, 2)) * time * (1 - time / td) * Math.exp(-time / tau)) / 60
		};
	});
	logger.debug('the is the accumulated glargine activity:%o', timeSinceGlargineAct);

	// compute the aggregated activity of last glargines in 27 hours

	let lastGlargines = timeSinceGlargineAct.filter(function (e) {
		return e.time <= 30;
	});
	logger.debug('these are the last glargines and activities:%o', lastGlargines);

	var resultGlaAct: number = lastGlargines.reduce(function (tot, arr) {
		return tot + arr.glargineActivity;
	}, 0);

	return resultGlaAct
	// logger.debug(resultGlaAct);

	// const fs = require('fs');
	// const { duration } = require('moment');
	// const GlaAct = JSON.stringify(resultGlaAct, null, 4);
	// fs.writeFile('./files/last_glargine_aggrACT.json', GlaAct, (err) => {
	//     if (err) {
	//         throw err;
	//     }
	//     logger.debug("aggregated GLA activity is now is saved as JSON.");
	//   });
};