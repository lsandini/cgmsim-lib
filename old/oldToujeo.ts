import logger from '../src/utils';

export default function (weight, toujeos) {
    // const weight = parseInt(process.env.WEIGHT);
    // var moment = require('moment');
    // const toujeos = require('./files/last_toujeo.json');
    var jsontou = JSON.stringify(toujeos);
    var toujeo_data = JSON.parse(jsontou);
    logger.debug(toujeo_data);

    // activities be expressed as U/min !!!
    let timeSinceToujeoAct = toujeo_data.map((entry) => {
        var time = entry.time;
        var dose = entry.dose;
        var duration = 24 + (14 * dose) / weight;
        var peak = duration / 2.5;
        var tp = peak;
        var td = duration;

        var tau = (tp * (1 - tp / td)) / (1 - (2 * tp) / td);
        var a = (2 * tau) / td;
        var S = 1 / (1 - a + (1 + a) * Math.exp(-td / tau));

        var toujeoActivity = 0;
        return {
            ...entry,
            time: time,
            toujeoActivity:
                (dose *
                    (S / Math.pow(tau, 2)) *
                    time *
                    (1 - time / td) *
                    Math.exp(-time / tau)) /
                60,
        };
    });
    logger.debug(
        'the is the accumulated toujeo activity:%o',
        timeSinceToujeoAct
    );

    // compute the aggregated activity of last toujeos in 27 hours

    let lastToujeos = timeSinceToujeoAct.filter(function (e) {
        return e.time <= 30;
    });
    logger.debug('these are the last toujeos and activities:%o', lastToujeos);

    var resultTouAct = lastToujeos.reduce(function (tot, arr) {
        return tot + arr.toujeoActivity;
    }, 0);

    logger.debug('result: %0', resultTouAct);
    return resultTouAct;
}
