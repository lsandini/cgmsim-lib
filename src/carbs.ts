import { Treatment } from './Types';
import logger, { getDeltaMinutes } from './utils';

//const logger = pino();

export default function carbs(treatments: Treatment[] = [], carbsAbs: number, isf: number, cr: number): number {
	const isfMMol = isf / 18; //(mmol/l)/U

	const meals = treatments
		.filter(e => e.carbs && getDeltaMinutes(e.created_at) <= 360)
		.map(e => ({
			...e,
			minutesAgo: getDeltaMinutes(e.created_at),
		}))
		.filter(e=>e.minutesAgo>=0);

	logger.debug('Last 6 hours meal: %o', meals);

	const carbs = meals || [];
	const carbAbsTime = carbsAbs; // meal absorption time in min default 360 or 6 hours
	const fast_carbAbsTime = carbAbsTime / 6; // = 1 h or 60 min
	const slow_carbAbsTime = carbAbsTime / 1.5; // = 4 h or 240 min

	const timeSinceMealAct = carbs.map(entry => {
		const minutesAgo = entry.minutesAgo;
		const carbs_g = entry.carbs;

		// the first 40g of every meal at most are always considered fast carbs ???
		// No, let the amount of fast carbs be totally random, but at most 30 g.
		const fast = Math.min(entry.carbs, 40);

		// the amount exceeding 30 grams will be randomly split into fast and slow carbs
		const rest = entry.carbs - fast;
		const FSR = (Math.random() * (0.4 - 0.1) + 0.1); // FSR = FAST RANDOM RATIO

		// all fast carbs counted together
		const fast_carbs = fast + (FSR * rest);

		// the remainder is slow carbs
		const slow_carbs = (1 - FSR) * rest;

		logger.debug('carbs_g:', carbs_g, 'fast:', fast, 'rest:', rest, 'fast_carbs:', fast_carbs, 'slow_carbs: %o', slow_carbs);

		let fast_carbrate = 0;
		let slow_carbrate = 0;


		if (minutesAgo < (fast_carbAbsTime / 2)) {
			const AT2 = Math.pow(fast_carbAbsTime, 2);
			fast_carbrate = (fast_carbs * 4 * minutesAgo) / AT2;
			//COB = (fast_carbs * 2 * Math.pow(t, 2)) / AT2;
		} else if (minutesAgo < (fast_carbAbsTime)) {
			fast_carbrate = (fast_carbs * 4 / fast_carbAbsTime) * (1 - (minutesAgo / fast_carbAbsTime));
			// const AAA = (4 * fast_carbs / fast_carbAbsTime);
			// const BBB = Math.pow(t, 2) / (2 * fast_carbAbsTime);
			// COB = (AAA * (t - BBB)) - fast_carbs;
		} else {
			fast_carbrate = 0;
			// COB = 0;
			logger.debug('fast carb absorption rate: %o', fast_carbrate);
		}

		if (minutesAgo < (slow_carbAbsTime / 2)) {
			const AT2 = Math.pow(slow_carbAbsTime, 2);
			slow_carbrate = (slow_carbs * 4 * minutesAgo) / AT2;
			//COB = (slow_carbs * 2 * Math.pow(t, 2)) / AT2;
		} else if (minutesAgo < (slow_carbAbsTime)) {
			slow_carbrate = (slow_carbs * 4 / slow_carbAbsTime) * (1 - (minutesAgo / slow_carbAbsTime));
			// const AAA = (4 * slow_carbs / slow_carbAbsTime);
			// const BBB = Math.pow(t, 2) / (2 * slow_carbAbsTime);
			//COB = (AAA * (t - BBB)) - slow_carbs;
		} else {
			slow_carbrate = 0;
			// COB = 0;
			logger.debug('slow carb absorption rate: %o', slow_carbrate);
		}

		return fast_carbrate + slow_carbrate;
	});
	logger.debug(timeSinceMealAct);


	const totalCarbRate = timeSinceMealAct.reduce((tot, carbrate) => tot + carbrate, 0);
	const DBGC = (isfMMol / cr) * totalCarbRate; //DeltaBloodGlucoseFromCarbs
	logger.debug(`CARB RATE:%o`, totalCarbRate);
	logger.info(`Delta blood Glucose From Carbs per minute:%o`, DBGC);

	return DBGC;
};