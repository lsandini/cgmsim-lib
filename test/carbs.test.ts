import moment = require('moment');
import carbs from '../src/carbs';
import { oldCarbs } from '../old/oldCarbs';
import { assert } from 'console';
const now = '2001-01-01T07:00:00';

beforeEach(() => {
	jest.useFakeTimers('modern');
	jest.setSystemTime(new Date(now));

})
afterAll(() => {
	jest.useRealTimers();
});

const mockMath = Object.create(global.Math);
mockMath.random = () => 0.5;
global.Math = mockMath;

const minutesAgo = (min) => moment(now).add(-min, 'minutes').toISOString();
describe('Carbs test', () => {

	test('test carbs without treatments return 0', () => {
		const r = carbs([], 360, 30, 10)
		expect(r).toBe(0);

	});
	test('test carbs with old treatments return 0', () => {
		const r = carbs([{
			carbs: 44, created_at: minutesAgo(361),
		}], 360, 30, 10)
		expect(r).toBe(0);
	});


	test('test carbs <=40 with active treatments return fix carbsActive', () => {
		const r = carbs([{
			carbs: 40, created_at: minutesAgo(45)
		}], 360, 30, 10)
		expect(r).toMatchSnapshot();
	});

	test('test carbs <=40 with active treatments return carbsActive', () => {
		const r = carbs([{
			carbs: 20, created_at: minutesAgo(1),
		}, {
			carbs: 20, created_at: minutesAgo(45),
		}], 360, 30, 10)
		expect(r).toMatchSnapshot();
	});

	test('carbs 40 min ago are more active then 60 min ago ', () => {
		const r40 = carbs([{
			carbs: 20, created_at: minutesAgo(40)
		}], 360, 30, 10)
		const r60 = carbs([{
			carbs: 20, created_at: minutesAgo(60)
		}], 360, 30, 10)
		expect(r40).toBeGreaterThan(r60);
	});

	test('carbs 5 min ago are less active then 40 min ago', () => {
		const r5 = carbs([{
			carbs: 20, created_at: minutesAgo(5)
		}], 360, 30, 10)
		const r60 = carbs([{
			carbs: 20, created_at: minutesAgo(40)
		}], 360, 30, 10)
		expect(r5).toBeLessThan(r60);
	});

	test('test carbs >40 with active treatments return carbsActive random', () => {
		const r = carbs([{
			carbs: 41, created_at: minutesAgo(1),
		}, {
			carbs: 41, created_at: minutesAgo(45),
		}], 360, 30, 10)
		expect(r).toBeGreaterThan(0);
	});

	it.each([
		[[2, 21]],
		[[3, 7, 21]],
		[[5, 27, 55, 98]],
		[[15, 22, 35, 83, 143]],
	])('test carb %p', (numbers: number[]) => {
		const e = numbers.map(n => ({
			carbs: 20, created_at: minutesAgo(n)
		}));
		const r = carbs(e, 360, 30, 10);
		expect(r).toMatchSnapshot();

	});
})


describe('Carbs test old compare', () => {
	beforeEach(() => {
		jest.useFakeTimers('modern');
		jest.setSystemTime(new Date(now));

	})
	afterAll(() => {
		jest.useRealTimers();
	});
	test('should first', () => {

		let now = moment();
		const oldC:number[] = [];
		const newC:number[] = [];
		const isf = 36;
		const cr = 10;
		// 
		for (let i = 0; i < 360; i++) {

			const treatment = [{
				carbs: 41, created_at: minutesAgo(1), time: moment(minutesAgo(1)).toDate().getTime()
			},];
			const carbAbs = 360;
			const newCarb = carbs(treatment, carbAbs, isf, cr);
			const old = oldCarbs(treatment, carbAbs)
			newC.push(newCarb)  // new raw value is multiplied by isfMMol/CR  or 2/10 or 1/5 !
			oldC.push(old);     // old raw value is not multiplied by CSF (isfMMol/CR)
			now = now.add(1, "minutes");
			jest.setSystemTime(now.toDate());
		}
		const newT = newC.reduce((acc, i) => ((i * (cr / (isf / 18))) + acc), 0)
		const oldT = oldC.reduce((acc, i) => i + acc, 0)
		expect(newT.toFixed(5)).toBe(oldT.toFixed(5));
		expect(newT.toFixed(5)).toBe(oldT.toFixed(5))
	})

});
