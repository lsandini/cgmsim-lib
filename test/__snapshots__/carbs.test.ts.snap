// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carbs test test carb [2, 21] 1`] = `0.07904806557343512`;

exports[`Carbs test test carb [3, 7, 21] 1`] = `0.08760171378902377`;

exports[`Carbs test test carb [5, 27, 55, 98] 1`] = `0.12409878722364139`;

exports[`Carbs test test carb [15, 22, 35, 83, 143] 1`] = `0.185559728442568`;

exports[`Carbs test test carbs <=40 with active treatments return carbsActive 1`] = `0.0362018370092477`;

exports[`Carbs test test carbs <=40 with active treatments return fix carbsActive 1`] = `0.07393044596670624`;
